File: python/src/engines/notation.py
"""Notation bridge between this project's chord spelling and JazzNet's.

This project spells flats with a trailing ``b`` (``Bb:maj7``, ``Db:7``); JazzNet's
vocabulary spells them with a dash (``B-:maj7``, ``D-:7``). Sharps (``F#``) and the
``Root:quality`` colon format are identical, so only the flat marker differs.

Applied around the neural engines: convert the incoming chord to JazzNet spelling
before lookup, and convert the model's reply back so the Max chord parser (which
expects ``Bb``) can sonify it. Non-chord tokens pass through unchanged.
"""

from __future__ import annotations


def to_jazznet(chord: str) -> str:
    """``Bb:maj7`` -> ``B-:maj7`` (flat ``b`` in the root becomes ``-``)."""
    if not chord or ":" not in chord:
        return chord.replace("b", "-") if chord else chord
    root, _, qual = chord.partition(":")
    return f"{root.replace('b', '-')}:{qual}"


def from_jazznet(chord: str) -> str:
    """``B-:maj7`` -> ``Bb:maj7`` (dash flat in the root becomes ``b``)."""
    if not chord or ":" not in chord:
        return chord.replace("-", "b") if chord else chord
    root, _, qual = chord.partition(":")
    return f"{root.replace('-', 'b')}:{qual}"

File: versions/v4-theory/max/engine/notation.js
"use strict";

/**
 * notation.js — spelling bridge between this project's chords and JazzNet's.
 *
 * We spell flats with a trailing 'b' (Bb:maj7); JazzNet spells them with '-'
 * (B-:maj7). Sharps and the Root:quality colon format are identical, so only
 * the flat marker in the ROOT differs. Direct port of
 * python/src/engines/notation.py — the neural engines convert in before lookup
 * and out before sonifying, so the transform must be exactly reversible for the
 * colon case.
 *
 * Note the no-colon branch mirrors the Python verbatim: it replaces every flat
 * marker in the whole token (there is no root/quality split without a ':').
 */

// Replace all occurrences (Python str.replace replaces every match).
function replaceAll(s, from, to) {
  return s.split(from).join(to);
}

/** toJazznet('Bb:maj7') -> 'B-:maj7' (flat 'b' in the root becomes '-'). */
function toJazznet(chord) {
  if (!chord || chord.indexOf(":") === -1) {
    return chord ? replaceAll(chord, "b", "-") : chord;
  }
  const i = chord.indexOf(":");
  const root = chord.slice(0, i);
  const qual = chord.slice(i + 1);
  return `${replaceAll(root, "b", "-")}:${qual}`;
}

/** fromJazznet('B-:maj7') -> 'Bb:maj7' (dash flat in the root becomes 'b'). */
function fromJazznet(chord) {
  if (!chord || chord.indexOf(":") === -1) {
    return chord ? replaceAll(chord, "-", "b") : chord;
  }
  const i = chord.indexOf(":");
  const root = chord.slice(0, i);
  const qual = chord.slice(i + 1);
  return `${replaceAll(root, "-", "b")}:${qual}`;
}

module.exports = { toJazznet, fromJazznet };
